design/20200326.extensible-certificate-controller.md

```diff
@@ -0,0 +1,388 @@
+          event and delete all but one.
+    * If no CertificateRequest for the current revision exists:
+        Create a CertificateRequest based on the `certificate.spec`.
+    * Verify the public key of the certificate request matches the public key
+      of the stored `status.nextPrivateKeySecretName`
+        * If not, delete the CertificateRequest & log an Event
+    * Verify the CSR options match what is requested in `certificate.spec`
+        * If not, delete the CertificateRequest & log an Event
+* If the `Issuing` condition is `False` or not set:
+    * Delete any CertificateRequest resources that do not have a 'revision'
+      equal to `status.revision`.
+
+### issuing
+
+This controller will copy the `status.certificate` and `status.ca` fields from
+valid CertificateRequest resources and the `tls.key` from the
+`status.nextPrivateKeySecretName` Secret resource into the `spec.secretName`.
+
+It is responsible for encoding the private key and certificate data into the
+appropriate format.
+
+Once the key-pair has been written to `spec.secretName`, it will set the
+`status.revision` field to that of the CertificateRequest and remove the
+`Issuing` status condition (in the same Update call).
+
+It will behave as follows:
+
+* If the 'Issuing' condition is `True`:
+    * If the `status.nextPrivateKeySecretName` field is not set:
+        * Do nothing
+    * Find the CertificateRequest for `status.revision` + 1
+        * If none exist, do nothing.
+        * If multiple exist, do nothing (`requestmanager` will handle this)
+    * If the CertificateRequest is not in a 'final' state, do nothing.
+    * If it is 'valid':
+        * Verify the public key of the certificate request matches the public key
+          of the stored `status.nextPrivateKeySecretName`
+            * If not, do nothing (`requestmanager` will handle this)
+        * Verify the CSR options match what is requested in `certificate.spec`
+            * If not, do nothing (`requestmanager` will handle this)
+        * Encode and issue the key-pair (store it in the Secret)
+        * Set `status.revision` to revision of the CertificateRequest
+        * Remove `Issuing` status condition
+        * Clear `status.lastFailureTime` (if set)
+        * Log an Event
+    * If not 'valid':
+        * Set `status.lastFailureTime` (if not equal to CertificateRequest)
+        * Set `Issuing` status condition to `False` with reason explaining when
+          the request will be retried
+* If the 'Issuing' condition is `False` or not set:
+    * Do nothing
+
+#### trigger
+
+The trigger plugin is responsible for observing the state of the currently
+issued `spec.secretName` and the rest of the `certificate.spec` fields to
+determine whether a re-issuance is required.
+It triggers re-issuance by adding the `Issuing` status condition when a new
+certificate is required.
+
+These conditions will cause a request to be triggered:
```

Comment:
What is the behavior on a secret owner mismatch? Most importantly does owner stay correct on for example a cert-manager re-install? I also had this note somewhere else in this doc but i lost it...